File: scratch/extract_data.js

const fs = require('fs');

function parseSql(filePath) {
    const content = fs.readFileSync(filePath, 'latin1');
    const mutasyabihat = [];
    const surahs = [];

    // Parse cekayatmutasyabihat
    const mutMatch = content.match(/INSERT INTO `cekayatmutasyabihat` .*? VALUES\s*([\s\S]*?);/);
    if (mutMatch) {
        const values = mutMatch[1];
        const rows = values.match(/\(\d+,\s*\d+,\s*\d+,\s*\d+\)/g);
        if (rows) {
            rows.forEach(r => {
                const parts = r.match(/\d+/g);
                mutasyabihat.append({
                    id: parseInt(parts[0]),
                    juz: parseInt(parts[1]),
                    surah: parseInt(parts[2]),
                    ayat: parseInt(parts[3])
                });
            });
        }
    }
    
    // Wait, regex above for mutasyabihat might be slow for many rows.
    // Let's use a simpler approach.
}

// Rewriting more efficiently
function parseEfficiently(filePath) {
    const content = fs.readFileSync(filePath, 'latin1');
    const mutasyabihat = [];
    const surahs = [];

    const lines = content.split('\n');
    let currentTable = null;

    for (let line of lines) {
        if (line.includes('INSERT INTO `cekayatmutasyabihat`')) {
            currentTable = 'mut';
            continue;
        } else if (line.includes('INSERT INTO `daftarsurah`')) {
            currentTable = 'surah';
            continue;
        } else if (line.trim().endsWith(';')) {
            if (currentTable) {
                // Process the last part of values
                processLine(line, currentTable, mutasyabihat, surahs);
            }
            currentTable = null;
            continue;
        }

        if (currentTable) {
            processLine(line, currentTable, mutasyabihat, surahs);
        }
    }

    return { mutasyabihat, surahs };
}

function processLine(line, table, mutasyabihat, surahs) {
    const matches = line.match(/\((.*?)\)/g);
    if (!matches) return;

    matches.forEach(m => {
        const parts = m.substring(1, m.length - 1).split(',');
        if (table === 'mut') {
            mutasyabihat.push({
                id: parseInt(parts[0].trim()),
                juz: parseInt(parts[1].trim()),
                surah: parseInt(parts[2].trim()),
                ayat: parseInt(parts[3].trim())
            });
        } else if (table === 'surah') {
            // (id, kategori, nosurat, nama, awal, akhir)
            surahs.push({
                id: parseInt(parts[0].trim()),
                juz: parseInt(parts[1].trim()),
                surah: parseInt(parts[2].trim()),
                nama: parts[3].trim().replace(/'/g, ""),
                awal: parseInt(parts[4].trim()),
                akhir: parseInt(parts[5].trim())
            });
        }
    });
}

const data = parseEfficiently('d:/emaqra+timer + rekap nilai/emaqra.sql');
fs.writeFileSync('d:/emaqra+timer + rekap nilai/parsed_data.json', JSON.stringify(data, null, 2));
console.log(`Extracted ${data.mutasyabihat.length} mutasyabihat and ${data.surahs.length} surah ranges.`);

File: scratch/extract_data.py

import re
import json

def parse_sql(file_path):
    mutasyabihat = []
    surahs = []
    
    with open(file_path, 'r', encoding='latin1') as f:
        content = f.read()
        
    # Parse cekayatmutasyabihat
    mut_match = re.search(r"INSERT INTO `cekayatmutasyabihat` .*? VALUES\s*(.*?);", content, re.DOTALL)
    if mut_match:
        values = mut_match.group(1)
        # (id, juz, nosurat, ayat)
        rows = re.findall(r"\((\d+),\s*(\d+),\s*(\d+),\s*(\d+)\)", values)
        for r in rows:
            mutasyabihat.append({
                'id': int(r[0]),
                'juz': int(r[1]),
                'surah': int(r[2]),
                'ayat': int(r[3])
            })
            
    # Parse daftarsurah
    surah_match = re.search(r"INSERT INTO `daftarsurah` .*? VALUES\s*(.*?);", content, re.DOTALL)
    if surah_match:
        values = surah_match.group(1)
        # (id, kategori, nosurat, nama, awal, akhir)
        rows = re.findall(r"\((\d+),\s*(\d+),\s*(\d+),\s*'([^']*)',\s*(\d+),\s*(\d+)\)", values)
        for r in rows:
            surahs.append({
                'id': int(r[0]),
                'juz': int(r[1]), # kategori seems to be juz
                'surah': int(r[2]),
                'nama': r[3],
                'awal': int(r[4]),
                'akhir': int(r[5])
            })
            
    return mutasyabihat, surahs

mutasyabihat, surahs = parse_sql('d:/emaqra+timer + rekap nilai/emaqra.sql')

data = {
    'mutasyabihat': mutasyabihat,
    'surahs': surahs
}

with open('d:/emaqra+timer + rekap nilai/parsed_data.json', 'w') as f:
    json.dump(data, f)

print(f"Extracted {len(mutasyabihat)} mutasyabihat and {len(surahs)} surah ranges.")
